### ISSUE TITLE
Animancer creates state "clones" of different types than the originals

### ISSUE THREAD

**Issue by davidrochin** [KybernetikGames/animancer#88]
# Environment

- Animancer Version Number: 4.4
- Animancer Pro or Lite: Pro
- Unity Version: 2019.4.16
- Platform: Windows

# Description

I'm storing and accessing a lot of metadata in my custom objects that extend from ClipTransition. I pass them to Animancer just like `_myLayer.Play(customClipTransition)`. The problem is that Animancer sometimes creates "copies" of my transitions to use them as states, and the copy is of a different type than the original, so I lose all my metadata.

I identified the line where this happens:
![image](https://user-images.githubusercontent.com/7517516/101570091-ec748600-3992-11eb-9dff-55ce1a7288a3.png)

I can see a comment up there that describes how my problem could be solved in the future, but I can't find any traces of this being fixed in newer versions of Animancer. I thought of storing my metadata in just the Key instead, but Animancer is not respecting that either (it just uses the clip as Key, so I still lose all my data).

Was this fixed? Any workarounds?

**Comment by KybernetikGames** [comment]
I haven't done any work on that area since no one else has mentioned it yet, but I think there might be some confusion about your terminology.

Animancer doesn't create copies of your transition, it creates copies of the state created by your transition. So unless you have also made a class that inherits from `ClipState` to implement a custom state type, it should still work fine.

If your transition has any additional data that it needs to set on the state, you should be doing that in your transition's `Apply` method which gets called by `animancer.Play(transition)` rather than your transition's `CreateState` method which only gets called to create the original state but isn't used for any duplicates since the original state doesn't have a reference to the transition that created it.

**Comment by davidrochin** [comment]
I see. I overridden the `Apply(AnimancerState)` method to set a custom `Key`, but I got a `An item with the same key has already been added` error. I guess it was expected, but now I have no idea of how to include all my metadata in my states. What would be the recommended way of doing this?

**Comment by KybernetikGames** [comment]
What metadata are you trying to include? Can you show me your code?

**Comment by davidrochin** [comment]
> What metadata are you trying to include? Can you show me your code?

Stuff like "Block character movement" and "Use root motion". Here is some chunks of my code. 

This is my custom transition scriptable object (the name is to make it more clear to designers who know nothing about Animancer):
![image](https://user-images.githubusercontent.com/7517516/101583018-5c393f80-3998-11eb-8759-8ada22bb218f.png)

This is an example of a system where I need to access the data (`CharacterAnimationState `inherits from `AnimationState` and just adds some fields):
![image](https://user-images.githubusercontent.com/7517516/101583435-bafeb900-3998-11eb-97f8-10d69868b50d.png)

I play my states like this:
![image](https://user-images.githubusercontent.com/7517516/101583560-f5685600-3998-11eb-8fd7-60d18b9273ff.png)

All for the sake of having this for my designers:
![image](https://user-images.githubusercontent.com/7517516/101583602-0c0ead00-3999-11eb-9ab2-6cfd9addcbaf.png)

**Comment by KybernetikGames** [comment]
That's interesting. Using the `Key` to reference data like that isn't something I've considered before. I would put a field like `Block Movement` in the state machine system rather than the animation system, but it should be possible to do it like that.

The clone states created by `GetOrCreateWeightlessState` each use the state they were cloned from as their `Key` so you could get back to the original `Key` with a simple loop:

```
public static object GetOriginalKey(AnimancerState state)
{
    var key = state.Key;
    while (key is AnimancerState keyState)
    {
        key = keyState.Key;
    }
    return key;
}
```

Then cast that to `CharacterAnimationState` or whatever.

Also, you could send `UseRootMotion` straight to the `Animator` in `Apply` (assuming that's what you're using it for):

    state.Root.Component.Animator.useRootMotion = UseRootMotion;

**Comment by davidrochin** [comment]
> The clone states created by `GetOrCreateWeightlessState` each use the state they were cloned from as their `Key` so you could get back to the original `Key` with a simple loop:

Is that something you changed after 4.4? Right now I'm getting the AnimationClip as the key:
![image](https://user-images.githubusercontent.com/7517516/101585390-ebe0ed00-399c-11eb-8d11-349cf7695e53.png)

**Comment by KybernetikGames** [comment]
I meant if you still override the `Key` in your transition.

The `Key` of the first state a transition creates will be its `AnimationClip` unless you assign a different one (or override the transition's `Key` property to determine what it gives to the state it creates).

So if you're using the `Key` as metadata like that, you should just override the transition's `Key` property to return itself.

**Comment by davidrochin** [comment]
Ah yes, I'm still overriding Key like `public override object Key => this;`. I just removed the `Apply `method because that caused an error.

**Comment by KybernetikGames** [comment]
What does your code look like now?

**Comment by davidrochin** [comment]
Exactly the same except for the Apply method, which I removed. Btw, this is where the state gets created (red). Unlike the part below (blue ) it doesn't look like the original state is being used as the key:
![image](https://user-images.githubusercontent.com/7517516/101586187-9c032580-399e-11eb-9a88-da268f21e57a.png)

**Comment by KybernetikGames** [comment]
Oh right, I see the problem now. If the clip isn't the original state's key then the first clone can use it.

The idea was that the clone would then basically work as the original if you call `animancer.Play(clip)` but in this case you actually have a good reason for not wanting any of them to use the clip.

Just remove the red one and I'll take it out of the next version.

Also, in case you weren't aware, I just released [Animancer v6.0](https://kybernetik.com.au/animancer/docs/changes/animancer-v6-0/) a few days ago so you might consider updating.

**Comment by KybernetikGames** [comment]
Removing that red code isn't quite right. Put this there instead:

```
// Use any earlier state that is weightless.
var keyState = state;
while (true)
{
    keyState = keyState.Key as AnimancerState;
    if (keyState == null)
    {
        break;
    }
    else if (keyState.Weight == 0)
    {
        state = keyState;
        goto GotWeightlessState;
    }
}
```

**Comment by KybernetikGames** [comment]
Oh, and the `goto` label before the comment at the end of the method:

```
GotWeightlessState:

// Make sure it is on this layer and at time 0.
```

**Comment by davidrochin** [comment]
Thanks! I'll try it out tomorrow and let you know how it went.

**Comment by davidrochin** [comment]
It seems everything is working great. Thanks for the great support!

**Comment by KybernetikGames** [comment]
Happy to help. I'll keep the issue open until I release the next update with the fix in it just in case anyone else encounters the same thing.

**Comment by KybernetikGames** [comment]
This fix is included in Animancer v6.1 which is now available on [itch.io](https://kybernetik.itch.io/animancer-pro) and should be approved on the [Asset Store](https://assetstore.unity.com/packages/tools/animation/animancer-pro-116514?aid=1100l8ah5&utm_source=aff) in a few days.
